apacheconfig/wloader.py

```diff
@@ -0,0 +1,198 @@
+        """Returns object typestring as defined by the apacheconfig parser.
+
+        Returns:
+            The typestring (as defined by the apacheconfig parser) for
+            this node. The possible values for this are ``"comment"``,
+            ``"statement"``, ``"include"``, ``"includeoptional"``.
+        """
+        return self._type
+
+    @property
+    def whitespace(self):
+        """Returns preceding whitespace for this node.
+
+        For example::
+
+            ItemNode('\\n  option value').whitespace => "\\n  "
+            ItemNode('option value').whitespace => ""
+            ItemNode('\\n  # comment').whitespace => "\\n  "
+        """
+        return self._whitespace
+
+    @whitespace.setter
+    def whitespace(self, value):
+        """See base class. Operates on preceding whitespace."""
+        self._whitespace = value
+
+    @staticmethod
+    def parse(raw_str, parser):
+        """Factory for :class:`apacheconfig.ItemNode` by parsing data from a
+        config string.
+
+        Args:
+            parser (:class:`apacheconfig.ApacheConfigParser`): specify the
+                parser to use. Can be created by ``native_apache_parser()``.
+        Returns:
+            :class:`apacheconfig.ItemNode` containing metadata parsed from
+            ``raw_str``.
+        """
+        raw = parser.parse(raw_str)
+        return ItemNode(raw[1])
+
+    @property
+    def name(self):
+        """Returns the name of this node.
+
+        The name is the first non-whitespace token in the directive. Cannot be
+        written. For comments, is the entire comment.
+        """
+        return self._raw[0]
+
+    @property
+    def has_value(self):
+        """Returns ``true`` if this :class:`apacheconfig.ItemNode` has a value.
+
+        ``ItemNode`` objects don't have to have a value, like option/value
+        directives with no value, or comments.
+        """
+        return len(self._raw) > 1
+
+    @property
+    def value(self):
+        """Returns the value of this item as a string.
+
+        The "value" is anything but the name. Can be overwritten.
+        """
+        if not self.has_value:
+            return
+        return self._raw[-1]
+
+    @value.setter
+    def value(self, value):
+        """Sets for the value of this item.
+
+        Args:
+            value (str): string to set new value to.
+
+          .. todo:: (sydneyli) convert `value` to quotedstring when quoted
+        """
+        if not self.has_value:
+            self._raw = self._raw + (" ", value,)
+        self._raw = self._raw[0:-1] + (value,)
+
+    def dump(self):
+        return (self.whitespace +
+                "".join([_restore_original(word) for word in self._raw]))
+
+    def __str__(self):
+        return ("%s(%s)"
+                % (self.__class__.__name__,
+                   str([self._type] +
+                       [_restore_original(word) for word in self._raw])))
+
+
+def native_apache_parser(options=None):
```

Comment:
It seems that this method mainly applies some specific options to lexer/parser/config objects. In some sense, the instantiation code is not relevant to this purpose, right?

So what if we introduce some sort of syntax "flavors" i.e. sets of options, perhaps as global constants, that the user can just pick or mix and match those options to specialize parsing suite to the desired syntax?

Reply:
Yes, I like the idea of having and exporting a set of common option configurations. Something like `NATIVE_APACHE_PARSER_OPTIONS`?

The motivation behind this was primarily to have a simpler interface for creating a native Apache parser-- would you prefer not to increase the surface area of exported functions to maintain? Either way is fine by me.

Reply:
> Yes, I like the idea of having and exporting a set of common option configurations. Something like `NATIVE_APACHE_PARSER_OPTIONS`?

Yeah... BTW, with current design the options are not only for the parser, but for lexer and loader as well.

> The motivation behind this was primarily to have a simpler interface for creating a native Apache parser-- would you prefer not to increase the surface area of exported functions to maintain? Either way is fine by me.

I'd try not to introduce something overly specific and not widely reusable whenever possible.

How about this:

    from apacheconfig import flavors
    from apachaconfig import make_loader

    with make_loader(**flavors.NATIVE_APACHE) as loader:
        ...

If we also need just loader, would it make sense to have a similar function? Or may be we can rename & parameterize `make_loader` so that it would become a universal factory function?

Either way, this factory function(s) would be driven by pre-defined and/or user-specified options.

Reply:
I'm a fan of exposing the option sets (`flavors`, as you put it) for the time being-- I have to think a bit more on what the factory function might look like, maybe once the other interfaces are in.

Reply:
I am thinking, that functionally we specialize parser to work on a specific kind (flavor) of grammar. Each flavor can be composed from a bunch of options.

Reply:
I gave it a shot with a `flavors.py`-- lmk what you think!